lib/Interpreter/CIFactory.cpp

```diff
@@ -338,9 +337,442 @@ static bool getVisualStudioDir(std::string &path) {
+    container_t m_Saved;
+
+  public:
+    
+    void addArgument(const char* arg, std::string value) {
+      m_Saved.push_back(std::make_pair(arg,std::move(value)));
+    }
+    container_t::const_iterator begin() const { return m_Saved.begin(); }
+    container_t::const_iterator end() const { return m_Saved.end(); }
+    bool empty() const { return m_Saved.empty(); }
+  };
+
+#ifndef _MSC_VER
+
+  static void CompilerIncludePaths(const char* Compiler,
+                                   llvm::SmallVectorImpl<char>& Buf,
+                                   AdditionalArguments& Args) {
+    std::string CppInclQuery("LC_ALL=C ");
+    CppInclQuery.append(Compiler);
+    CppInclQuery.append(" -xc++ -E -v /dev/null 2>&1 >/dev/null "
+                        "| awk '/^#include </,/^End of search"
+                        "/{if (!/^#include </ && !/^End of search/){ print }}' "
+                        "| GREP_OPTIONS= grep -E \"(c|g)\\+\\+\"");
+
+    if (FILE *pf = ::popen(CppInclQuery.c_str(), "r")) {
+      Buf.resize(Buf.capacity_in_bytes());
+      while (fgets(&Buf[0], Buf.capacity_in_bytes(), pf) && Buf[0]) {
+        size_t lenbuf = strlen(&Buf[0]);
+        Buf[lenbuf - 1] = 0; // remove trailing \n
+        // Skip leading whitespace:
+        const char *start = &Buf[0];
+        while (start < (&Buf[0] + lenbuf) && *start == ' ')
+          ++start;
+        if (*start) {
+          if (llvm::sys::fs::is_directory(start))
+            Args.addArgument("-I", start);
+        }
+      }
+      ::pclose(pf);
+    } else
+      llvm::errs() << "popen failed for '" << CppInclQuery << "'\n";
+
+    // Return the query in Buf on failure
+    if (Args.empty()) {
+      Buf.resize(0);
+      Buf.insert(Buf.begin(), CppInclQuery.begin(), CppInclQuery.end());
+    }
+  }
+
+  static bool AddCxxPaths(llvm::StringRef PathStr, AdditionalArguments& Args) {
+    bool Success = true;
+    llvm::SmallVector<llvm::StringRef, 6> Paths;
+    for (std::pair<llvm::StringRef, llvm::StringRef> Split
+           = PathStr.split(':');
+         !Split.second.empty(); Split = PathStr.split(':')) {
+      if (!llvm::sys::fs::is_directory(Split.first)) {
+        Success = false;
+        break;
+      }
+      Paths.push_back(Split.first);
+      PathStr = Split.second;
+    }
+    // Add remaining part
+    if (Success && llvm::sys::fs::is_directory(PathStr)) {
+      for (llvm::StringRef Path : Paths)
+        Args.addArgument("-I", Path.str());
+      Args.addArgument("-I", PathStr.str());
+      return true;
+    }
+    return false;
+  }
+
+#endif
+  
+  ///\brief Adds standard library -I used by whatever compiler is found in PATH.
+  static void AddHostArguments(std::vector<const char*>& args,
+                               const char* llvmdir, const CompilerOpts& opts) {
+    static AdditionalArguments sArguments;
+    if (sArguments.empty()) {
+#ifdef _MSC_VER
+      // Honor %INCLUDE%. It should know essential search paths with vcvarsall.bat.
```

Comment:
Looks OK with me, the _MSC_VER part looks very similar to what is already there (unless I overlooked something)...